services/blockstorage/block_sync.go

```diff
@@ -0,0 +1,323 @@
+				requestBlocksTrigger.Reset(b.config.BlockSyncCollectResponseTimeout())
+			}
+		}
+
+		if state == BLOCK_SYNC_PETITIONER_ASK_FOR_BLOCKS {
+			requestBlocksTrigger.Stop()
+
+			count := len(blockAvailabilityResponses)
+
+			if count == 0 {
+				state = BLOCK_SYNC_STATE_START_SYNC
+				continue
+			}
+
+			b.reporting.Info("collected block availability responses", log.Int("num-responses", count))
+
+			// TODO in the future we might want to have a more sophisticated select function than that
+			syncSource := blockAvailabilityResponses[rand.Intn(count)]
+			syncSourceKey := syncSource.Sender.SenderPublicKey()
+
+			err := b.PetitionerSendBlockSyncRequest(gossipmessages.BLOCK_TYPE_BLOCK_PAIR, syncSourceKey)
+			if err != nil {
+				b.reporting.Info("could not request block chunk from source", log.Error(err), log.Stringable("source", syncSource.Sender))
+				state = BLOCK_SYNC_STATE_START_SYNC
+				continue
+			} else {
+				b.reporting.Info("requested block chunk from source", log.Stringable("source", syncSource.Sender))
+				state = BLOCK_SYNC_PETITIONER_WAITING_FOR_CHUNK
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case event = <-b.Events:
+			continue
+		}
+	}
+}
+
+func (b *BlockSync) dispatchEvent(state blockSyncState, event interface{}, blockAvailabilityResponses []*gossipmessages.BlockAvailabilityResponseMessage) (blockSyncState, []*gossipmessages.BlockAvailabilityResponseMessage) {
+	switch event.(type) {
+	case *gossipmessages.BlockAvailabilityRequestMessage:
+		message := event.(*gossipmessages.BlockAvailabilityRequestMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.SourceHandleBlockAvailabilityRequest(message)
+		}
+	case *gossipmessages.BlockAvailabilityResponseMessage:
+		message := event.(*gossipmessages.BlockAvailabilityResponseMessage)
+
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			err := b.PetitionerHandleBlockAvailabilityResponse(message)
+
+			if err != nil {
+				b.reporting.Info("received bad block availability response", log.Error(err))
+			} else {
+				blockAvailabilityResponses = append(blockAvailabilityResponses, message)
+			}
+		}
+	case *gossipmessages.BlockSyncRequestMessage:
+		message := event.(*gossipmessages.BlockSyncRequestMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.SourceHandleBlockSyncRequest(message)
+			return BLOCK_SYNC_STATE_IDLE, blockAvailabilityResponses
+		}
+	case *gossipmessages.BlockSyncResponseMessage:
+		message := event.(*gossipmessages.BlockSyncResponseMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.PetitionerHandleBlockSyncResponse(message)
+			return BLOCK_SYNC_STATE_IDLE, blockAvailabilityResponses
+		}
+	}
+
+	return state, blockAvailabilityResponses
+}
+
+func (b *BlockSync) PetitionerBroadcastBlockAvailabilityRequest() error {
```

Comment:
This method should be private, if it needs to be public for the tests it means that this is a smell and these tests should be converted to unit tests. If it's too complicated to change now, add a TODO and we'll do it once we introduce unit tests

Reply:
Fixed